pkg/register/component.go

```diff
@@ -0,0 +1,227 @@
+		err = c.keycloakClient.ExecuteActionsEmail(accessToken, "master", userID, *realmConf.RegisterRequiredActions)
+		if err != nil {
+			c.logger.Warn(ctx, "msg", "ExecuteActionsEmail failed", "err", err.Error())
+			return "", err
+		}
+	}
+
+	// Commit the DB transaction
+	err = tx.Commit()
+	if err != nil {
+		c.logger.Warn(ctx, "msg", "Can't commit transaction", "err", err.Error())
+		return "", errorhandler.CreateInternalServerError("database")
+	}
+
+	// store the API call into the DB
+	c.reportEvent(ctx, "REGISTER_USER", database.CtEventRealmName, realmName, database.CtEventUserID, userID)
+
+	return username, nil
+}
+
+func (c *component) checkExistingUser(ctx context.Context, accessToken, realmName string, user apiregister.User) error {
+	// Email already used?
+	var existingUser, err = c.usersDBModule.GetUser(ctx, realmName, nil, nil, user.EmailAddress)
+	if err != nil {
+		c.logger.Warn(ctx, "msg", "Can't get user from db", "err", err.Error())
+		return errorhandler.CreateInternalServerError("database")
+	}
+	if existingUser == nil {
+		// Free to go on processing this user creation
+		return nil
+	}
+
+	var kcUser kc.UserRepresentation
+	kcUser, err = c.keycloakClient.GetUser(accessToken, realmName, *existingUser.UserID)
+	if err != nil {
+		c.logger.Warn(ctx, "msg", "Can't get user from keycloak", "err", err.Error())
+		return errorhandler.CreateInternalServerError("keycloak")
+	}
+
+	if kcUser.Enabled == nil || *kcUser.Enabled {
+		c.logger.Warn(ctx, "msg", "Attempt to register a user with email of an already enabled user")
+		// Should not leak that email is already in use
+		return errorhandler.CreateBadRequestError(errorhandler.MsgErrInvalidParam + ".user_emailAddress")
+	}
+
+	err = c.keycloakClient.DeleteUser(accessToken, realmName, *existingUser.UserID)
+	if err != nil {
+		c.logger.Warn(ctx, "msg", "Can't delete user from keycloak", "err", err.Error())
+		return errorhandler.CreateInternalServerError("keycloak")
+	}
+
+	err = c.usersDBModule.DeleteUser(ctx, realmName, *existingUser.UserID)
+	if err != nil {
+		c.logger.Warn(ctx, "msg", "Can't delete user from database", "err", err.Error(), "uid", *existingUser.UserID)
+		return errorhandler.CreateInternalServerError("database")
+	}
+
+	// Free to go on processing this user creation
+	return nil
+}
+
+func (c *component) generateUnusedUsername(ctx context.Context, realmName string) (string, error) {
+	for {
+		var username = c.generateUsername("0123456789", 8)
+		var dbUser, err = c.usersDBModule.GetUser(ctx, realmName, nil, &username, nil)
+		if err != nil {
+			c.logger.Warn(ctx, "msg", "Can't get user from database", "err", err.Error())
+			return "", errorhandler.CreateInternalServerError("database")
+		}
+		if dbUser == nil {
+			return username, nil
+		}
+	}
+}
+
+func (c *component) generateUsername(alphabet string, length int) string {
+	var chars = []rune(alphabet)
+	var b strings.Builder
+
+	for j := 0; j < length; j++ {
+		b.WriteRune(chars[rand.Intn(len(chars))])
```

Comment:
It is maybe not a security critical, but maybe we could use crypto/rand instead of math/rand ?

Reply:
It seems we already use math/rand everywhere else and crypto/rand works with a io.Reader... so I guess it is simpler to keep using math/rand